File: pyos/paging/lru.py
def findLRU(time, n):
    min_time = time[0]
    pos = 0
    for i in range(1, n):
        if time[i] < min_time:
            min_time = time[i]
            pos = i
    return pos


def main():
    num_frames = int(input("Enter number of frames: "))
    num_pages = int(input("Enter number of pages: "))
    
    print("Enter page reference string: ")
    pages = [int(input()) for _ in range(num_pages)]

    frames = [-1] * num_frames  # Initialize frames as empty
    time = [0] * num_frames  # Track the time each frame was last used
    page_faults = 0
    counter = 0

    for i in range(num_pages):
        found = False

        for j in range(num_frames):
            if frames[j] == pages[i]:
                found = True
                counter += 1
                time[j] = counter  # Update the last accessed time
                break

        if not found:  # Page fault occurs
            if -1 in frames:  # If there's an empty frame
                empty_index = frames.index(-1)
                frames[empty_index] = pages[i]
                counter += 1
                time[empty_index] = counter
            else:
                pos = findLRU(time, num_frames)
                frames[pos] = pages[i]
                counter += 1
                time[pos] = counter
            page_faults += 1

        print("\nFrames: ", end="")
        for frame in frames:
            print(frame, end=" ")

    print(f"\nTotal Page Faults = {page_faults}")


if __name__ == "__main__":
    main()

File: Page Replacement/LRU.c
#include <stdio.h>

int findLRU(int time[], int n) {
    int i, minimum = time[0], pos = 0;
    for (i = 1; i < n; ++i) {
        if (time[i] < minimum) {
            minimum = time[i];
            pos = i;
        }
    }
    return pos;
}

int main() {
    int frames[10], pages[30], time[10];
    int num_frames, num_pages, i, j, pos, page_faults = 0, counter = 0;

    printf("Enter number of frames: ");
    scanf("%d", &num_frames);

    printf("Enter number of pages: ");
    scanf("%d", &num_pages);

    printf("Enter page reference string: ");
    for (i = 0; i < num_pages; i++)
        scanf("%d", &pages[i]);

    for (i = 0; i < num_frames; i++) {
        frames[i] = -1;  // initialize frames as empty
        time[i] = 0;     // initialize time array
    }

    printf("\nPage Reference | Frames");
    printf("\n---------------|------------------");

    for (i = 0; i < num_pages; i++) {
        int found = 0;

        // Check if the page is already in the frame
        for (j = 0; j < num_frames; j++) {
            if (frames[j] == pages[i]) {
                found = 1;
                time[j] = ++counter;  // Update usage time
                break;
            }
        }

        if (!found) {  // Page fault occurs
            if (frames[j] == -1) {
                pos = j;  // First empty frame
            } else {
                pos = findLRU(time, num_frames);  // Find the least recently used
            }
            frames[pos] = pages[i];
            time[pos] = ++counter;
            page_faults++;
        }

        // Print the table row
        printf("\n%12d |", pages[i]);
        for (j = 0; j < num_frames; j++) {
            if (frames[j] != -1)
                printf(" %d", frames[j]);
            else
                printf(" -");
        }
    }

    printf("\n\nTotal Page Faults = %d\n", page_faults);
    return 0;
}
